general/development/tools/phpunit/upgrading-11.md

```diff
@@ -0,0 +1,198 @@
+---
+title: PHPUnit 11 Upgrade
+tags:
+- Testing
+- PHPUnit
+---
+
+<Since version="5.0" issueNumber="MDL-83468" />
+
+Moodle 5.0 updated the version of PHPUnit from version 9.6 to version 11.
+
+This guide is intended to help you update your tests to work with the newer versions.
+
+## Why did we make this change?
+
+PHPUnit 9.6 is currently in maintenance support, but we don't know for how much longer.
+
+We ultimately _need_ to update to a newer version of PHPUnit at some point to ensure that PHP version incompatibilities are addressed and we are able to keep testing Moodle properly.
+
+The most appropriate target is for Moodle 5.0 which follows our new Series naming and versioning policies.
+
+### Why not PHPUnit 10?
+
+- PHPUnit 10 is the next major version of PHPUnit and introduces a number of compatibility issues with PHPUnit 9.6.
+- PHPUnit 11 is the latest major version of PHPUnit and is _more_ backwards compatible with PHPunit 9.6 than PHPUnit 10 is.
+
+Many of the features removed in PHPUnit 10 were then partially added back in PHPUnit 11.
+
+To give a specific example, PHPUnit 9.6 has a `--debug` flag. This was removed in 10.0, and added back in 11.0 as an alias for a combination of other flags.
+
+There are lots of other examples of this kind of thing.
+
+## Can I make my tests work in both 9.6 and 11.4?
+
+Yes, _usually_. But it may require some changes to your tests and/or code. For the most part any such changes are a good thing and in all cases I've encountered they've been because the tests were wrong in the first place anyway.
+
+### I still can't make my tests work in both 9.6 and 11.4
+
+Unfortunately there is only so much we can do about this -- we can't support every possible combination of PHPUnit versions and we are limited by the versions of PHPUnit that are available in the PHP ecosystem for the versions of PHP that we support.
+
+One of the changes to Moodle's versioning policy was intended to make a balance between core maintainability, and plugin maintainability. Whilst we always strive to avoid making a breaking change to plugins, and to provide a stable for all _currently_ supported versions of Moodle, sometimes this is not possible.
+
+We do recommend that you create a new branch of your plugin for _each series release of Moodle_. That is recommend you have a branch for Moodle 4.x, and separate branch for Moodle 5.x.
+
+## Help -- my tests are broken!
+
+Generally speaking we have already tried to minimise the impact of this change by updating to PHPUnit 9.6 on all stables, which already had some support for some of the new 11.4 features.
+
+We've documented all of the issues we encountered in this transition below.
+
+### Empty data providers
+
+#### Problem
+
+Data providers _must_ now provide data. You cannot have data providers which return an empty Iterator/Array.
+
+```php
+/**
+ * @requires extension Redis
+ */
+```
+
+#### Solution - use the `@requires` annotation
```

Comment:
It's not quite clear to me how using `@requires` works or how to use it to fix the empty data provider issue. Is it being added to PHPDoc block of the test function that requires the data provider?

Maybe adding an example of an empty data provider with the corresponding `@requires` would be good?

Reply:
There's literally an example just above. I'll move it to the solution. We can't do _everything_. People need to take some responsibility and accountability for themselves.